crates/jstzd/src/task/task.rs

```diff
@@ -0,0 +1,253 @@
+use anyhow;
+use async_dropper::AsyncDrop;
+use tokio::task::JoinHandle as tokioJoinHandle;
+
+use async_trait::async_trait;
+use futures_util::{
+    task::{Context, Poll},
+    Future,
+};
+
+use std::pin::Pin;
+use tokio::signal::unix::{Signal, SignalKind};
+use tokio::sync::{mpsc, OnceCell};
+
+#[async_trait]
+pub trait Task {
+    /// A set of tasks which are all executed on the same 'execution backend'
+    type TaskSet;
+
+    /// Performs the unit of work for the task
+    async fn run_task(
+        self,
+        signals: mpsc::Receiver<Signal>,
+        task_group: &Self::TaskSet,
+    ) -> anyhow::Result<()>;
+
+    /// Spawns the task
+    async fn spawn_task(self) -> JoinHandle;
+}
+
+pub struct JoinHandle {
+    task: OnceCell<tokioJoinHandle<()>>,
```

Comment:
Would `Option` work instead? I don't see anywhere that we're relying on `OnceCell::set`?